plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java

```diff
@@ -366,20 +373,35 @@ public ColumnMetadata getColumnMetadata(
     public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
     {
         log.debug("listTableColumns(session=%s, prefix=%s)", session, prefix);
-        ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> columns = ImmutableMap.builder();
         List<SchemaTableName> tables = prefix.toOptionalSchemaTableName()
                 .<List<SchemaTableName>>map(ImmutableList::of)
                 .orElseGet(() -> listTables(session, prefix.getSchema()));
-        for (SchemaTableName tableName : tables) {
-            try {
-                Optional.ofNullable(getTableHandleIgnoringConflicts(session, tableName))
-                        .ifPresent(tableHandle -> columns.put(tableName, getTableMetadata(session, tableHandle).getColumns()));
-            }
-            catch (TableNotFoundException e) {
-                // table disappeared during listing operation
-            }
+
+        List<BigQueryTableHandle> tableHandles = processInParallel(tables, table -> getTableHandleIgnoringConflicts(session, table)).stream()
+                .filter(Objects::nonNull)
+                .map(BigQueryTableHandle.class::cast)
+                .collect(toImmutableList());
+
+        return processInParallel(tableHandles, tableHandle -> getTableMetadata(session, tableHandle)).stream()
+                .collect(toImmutableMap(ConnectorTableMetadata::getTable, ConnectorTableMetadata::getColumns));
+    }
+
+    protected <T, R> List<R> processInParallel(List<T> list, Function<T, R> function)
```

Comment:
There are two comments that came to my mind:
 - in case of high number of entities we don't want spin off high number of threads. Can you please limit this to for example 10 threads?
 - in case of a failure for any of elements we don't want to enumerate all entities but fail immediately. And that requires cancellation of the ones that are in progress.
 
 Both of that case requires some unit tests.

Reply:
Also do we care about ordering?

Reply:
@kokosing We don't care about ordering. We care about the rest of your concerns

Reply:
I've improved this PR.